src/instructlab/config.py

```diff
@@ -190,3 +192,48 @@ def write_config(cfg, config_file=DEFAULT_CONFIG):
 def get_api_base(host_port):
     """Returns server API URL, based on the provided host_port"""
     return f"http://{host_port}/v1"
+
+
+class _ColorFormatter(logging.Formatter):
+    """Colorize log lines"""
+
+    styles = {
+        logging.CRITICAL: {"fg": "bright_red"},
+        logging.ERROR: {"fg": "red"},
+        logging.WARNING: {"fg": "yellow"},
+        # logging.INFO: {},
+        logging.DEBUG: {"fg": "blue"},
+    }
+
+    def formatMessage(self, record: logging.LogRecord) -> str:
+        msg = super().formatMessage(record)
+        kwargs = self.styles.get(record.levelno)
+        if kwargs:
+            return click.style(msg, **kwargs)
+        return msg
+
+
+def configure_logging(
+    *,
+    level: Optional[str] = None,
+    fmt: str = "%(levelname)s %(asctime)s %(filename)s:%(lineno)d %(message)s",
+) -> None:
+    """Configure logging framework"""
+    root = logging.getLogger()
+
+    # reset handlers, removes existing stream logger
+    for handler in root.handlers[:]:
+        root.removeHandler(handler)
+        handler.close()
+
+    stream = logging.StreamHandler()
+    stream.setFormatter(_ColorFormatter(fmt=fmt))
+    root.addHandler(stream)
+
+    if level is None:
```

Comment:
Is this possible given we pass `ctx.obj.config.general.log_level` and that `config.py` as a validator function for the log level?

Reply:
`config.py` defines `log_level: Optional[StrictStr] = "INFO"`. That means the value of `log_level` is either None or a strict string. 

I think that the model should not use `Optional` here and in several other places. `Optional[T]` is syntactic sugar for `typing.Union[None, T]`. It does not mean optional argument.

https://docs.python.org/3/library/typing.html#typing.Optional
> Optional[X] is equivalent to X | None (or Union[X, None]).
> 
> Note that this is not the same concept as an optional argument, which is one that has a default. An optional argument with a default does not require the Optional qualifier on its type annotation just because it is optional.

Reply:
Thanks for the detailed explanation!

Reply:
Thanks for the feedback! It's much appreciated.

PR #1158 addresses the optional parameters in our pydantic model.